bus-mapping/src/evm/opcodes/extcodehash.rs

```diff
@@ -0,0 +1,295 @@
+use super::Opcode;
+use crate::{
+    circuit_input_builder::CircuitInputStateRef,
+    operation::{
+        AccountField, AccountOp, CallContextField, CallContextOp, TxAccessListAccountOp, RW,
+    },
+    state_db::Account,
+    Error,
+};
+use eth_types::{evm_types::GasCost, GethExecStep, ToAddress, ToWord};
+
+#[derive(Debug, Copy, Clone)]
+pub(crate) struct Extcodehash;
+
+impl Opcode for Extcodehash {
+    fn gen_associated_ops(
+        state: &mut CircuitInputStateRef,
+        steps: &[GethExecStep],
+    ) -> Result<(), Error> {
+        let step = &steps[0];
+        let stack_address = step.stack.last_filled();
+
+        // Pop external address off stack
+        let external_address = step.stack.last()?.to_address();
+        state.push_stack_op(RW::READ, stack_address, external_address.to_word())?;
+
+        // Read transaction id from call context
+        state.push_op(
+            RW::READ,
+            CallContextOp {
+                call_id: state.call()?.call_id,
+                field: CallContextField::TxId,
+                value: state.tx_ctx.id().into(),
+            },
+        );
+
+        // Update transaction access list for external_address
+        let account_previously_accessed = match step.gas_cost {
+            GasCost::WARM_STORAGE_READ_COST => true,
+            GasCost::COLD_ACCOUNT_ACCESS_COST => false,
+            _ => unreachable!(),
+        };
+        state.push_op(
```

Comment:
The step that does EXTCODEHASH is susceptible to be reverted (if its call or a parent call reverts).  On such revert, the changes on the access list must be reverted.
For all the operations that can be reverted we have a special method that must be used instead of `push_op`, which is `push_op_revertible`.  This method accumulates all the ops that will be reverted, and also sets the state write counter of the call.  See for example (this is in a WIP PR by Han):
https://github.com/han0110/zkevm-circuits/blob/8e8a355ad83a08cae8716b726fa9ff31d5ab54a4/bus-mapping/src/evm/opcodes/call.rs#L78-L86

Reply:
I've made this `push_op_revertible`, but I'm confused about why this is needed only for the write to the access list and not for the write to the stack. Is it because the stack is "automatically" cleared in the case of a revert, but the access list is not?

Reply:
I had to add `state_write_counter: Delta(1.expr()),` to the `StepStateTransition` in the gadget. Maybe we can rename `state_write_counter` -> `reversible_write_counter` if there are non-state writes that need to be included?

Reply:
> I've made this `push_op_revertible`, but I'm confused about why this is needed only for the write to the access list and not for the write to the stack. Is it because the stack is "automatically" cleared in the case of a revert, but the access list is not?

Exactly.  From all the operations in the rw_table there are three kinds:
- a. Those that can persist from one CALL to the next (TxAccessListAccount, TxAccessListAccountStorage, TxRefund, Logs)
- b. Those that can persist from one tx to the next (Account, AccountDestructed, AccountStorage)
- c. Those that only affect the context of the call (Stack, Memory, CallContext)

Stack, Memory and CallContext (case c) are local to a CALL, and are values that each call has and disappear when the call ends, so they are not tracked for possible reversion.

For case a and b, since the changes persist after the CALL finishes, we need to track the operations so that if they are executed in a non persistent call, they are eventually reverted.  (There's an exception here which is Logs, which we just don't generate when the step that creates them is not persistent.  We can do this because Logs don't have side effects that can be observable from within the EVM).

Reply:
> I had to add `state_write_counter: Delta(1.expr()),` to the `StepStateTransition` in the gadget. 

Sounds good!

> Maybe we can rename `state_write_counter` -> `reversible_write_counter` if there are non-state writes that need to be included?

Naming is hard :P I also think that using the word "state" in the context of the EVM is sometimes ambiguous, because it can refer to the State Trie (which defines the Ethereum Blockchain State), but it can also be used as the current state of execution when we model the EVM as a state machine (and this includes stack, memory, etc.).

The name `reversible_write_counter` sounds good to me.  @han0110 what do you think?

Reply:
Yeah `reversible_write_counter` makes sense to me.

(In my mind I would recognize `step_state` as "state of EVM as a state machine", and `state` as "state of Ethereum", but I believe we must have mixed the usage somewhere..)